### ISSUE TITLE
Crash: Restart Game Multithreading problem

### ISSUE THREAD

**Issue by Sumxx** [cocos2d/cocos2d-x#17352]
- cocos2d-x version: 3.6 and new
- devices test on:
- developing environments
   - NDK version:
   - Xcode version:
   - VS version:
   - browser type and version:

Steps to Reproduce:

Please look at those following key points:
```CPP
void HttpClient::sendImmediate(HttpRequest* request)
{
    if(!request)
    {
        return;
    }
    request->retain();
    // Create a HttpResponse object, the default setting is http access failed
    HttpResponse *response = new (std::nothrow) HttpResponse(request);
    auto t = std::thread(&HttpClient::networkThreadAlone, this, request, response);
    t.detach();
}
```
When running 'restart()', the JS engine will restart, because of the uneffective management of 't'.
```CPP
void Director::restartDirector()
{
    reset();
    // Texture cache need to be reinitialized
    initTextureCache();
    // Reschedule for action manager
    getScheduler()->scheduleUpdate(getActionManager(), Scheduler::PRIORITY_SYSTEM, false);
    // release the objects
    PoolManager::getInstance()->getCurrentPool()->clear();
    // Real restart in script level
#if CC_ENABLE_SCRIPT_BINDING
    ScriptEvent scriptEvent(kRestartGame, NULL);
    ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&scriptEvent);
#endif
}
```
## Crash point
Then the thread is waiting for the return of 'processResponse()', but, on the same time, the game has already restarted, and the running thread is out of control.

```CPP
void HttpClient::networkThreadAlone(HttpRequest* request, HttpResponse* response)
{
    char errorBuffer[ERROR_SIZE] = { 0 };
    processResponse(response, errorBuffer);
    auto scheduler = Director::getInstance()->getScheduler();
    scheduler->performFunctionInCocosThread([response, request]{
        const ccHttpRequestCallback& callback = request->getCallback();
        Ref* pTarget = request->getTarget();
        SEL_HttpResponse pSelector = request->getSelector();
        if (callback != nullptr)
        {
            callback(s_HttpClient, response);
        }
        else if (pTarget && pSelector)
        {
            (pTarget->*pSelector)(s_HttpClient, response);
        }
        response->release();
        // do not release in other thread
        request->release();
    });
}
```
When the network is relaticely poor, thread will cause that 'SAFE_REMOVE_OBJECT' release crash after the game 'restart()'.

```CPP
MinXmlHttpRequest::~MinXmlHttpRequest()
{
#define SAFE_REMOVE_OBJECT(callback) if (callback != NULL)\
    {\
        JS::RemoveObjectRoot(_cx, &callback);\
    }
    SAFE_REMOVE_OBJECT(_onreadystateCallback);
    SAFE_REMOVE_OBJECT(_onloadstartCallback);
    SAFE_REMOVE_OBJECT(_onloadendCallback);
    SAFE_REMOVE_OBJECT(_onloadCallback);
    SAFE_REMOVE_OBJECT(_onerrorCallback);
    SAFE_REMOVE_OBJECT(_onabortCallback);
    SAFE_REMOVE_OBJECT(_ontimeoutCallback);
    if (_httpRequest)
    {
        // We don't need to release _httpRequest here since it will be released in the http callback.
//        _httpRequest->release();
    }
    CC_SAFE_FREE(_data);
    CC_SAFE_RELEASE_NULL(_scheduler);
}
```
After all, should all the thread be managed and be closed effctively?

**Comment by Kingwl** [comment]
cooooooooooooooool!

**Comment by Sumxx** [comment]
@minggo please at

**Comment by minggo** [comment]
the restart() function is added to for JSB to just re-load js-script, please don't use it in multi thread environment.

**Comment by Sumxx** [comment]
Firstly of all, I don't use the Multi-Threading environment, and it's the part of network request to use it. Then if I want to solve the problem, I have to modify your network layer. Therefore, do you have any better  solution for the problem? HttpClient is cocos network function, not mine

**Comment by minggo** [comment]
Could you please provide a test case?

**Comment by Sumxx** [comment]
All right, please wait for a moment.

**Comment by Sumxx** [comment]
```js

var HelloWorldLayer = cc.Layer.extend({
    sprite:null,
    ctor:function () {
        //////////////////////////////
        // 1. super init first
        this._super();

        /////////////////////////////
        // 2. add a menu item with "X" image, which is clicked to quit the program
        //    you may modify it.
        // ask the window size
        var size = cc.winSize;

        // add a "close" icon to exit the progress. it's an autorelease object
        var closeItem = new cc.MenuItemImage(
            res.CloseNormal_png,
            res.CloseSelected_png,
            function () {
                cc.director.restart();
            }, this);
        closeItem.attr({
            x: size.width - 20,
            y: 20,
            anchorX: 0.5,
            anchorY: 0.5
        });

        var menu = new cc.Menu(closeItem);
        menu.x = 0;
        menu.y = 0;
        this.addChild(menu, 1);

        /////////////////////////////
        // 3. add your codes below...
        // add a label shows "Hello World"
        // create and initialize a label
        var helloLabel = new cc.LabelTTF("Hello World", "Arial", 38);
        // position the label on the center of the screen
        helloLabel.x = size.width / 2;
        helloLabel.y = 0;
        // add the label as a child to this layer
        this.addChild(helloLabel, 5);

        // add "HelloWorld" splash screen"
        this.sprite = new cc.Sprite(res.HelloWorld_png);
        this.sprite.attr({
            x: size.width / 2,
            y: size.height / 2,
            scale: 0.5,
            rotation: 180
        });
        this.addChild(this.sprite, 0);

        this.sprite.runAction(
            cc.sequence(
                cc.rotateTo(2, 0),
                cc.scaleTo(2, 1, 1)
            )
        );
        helloLabel.runAction(
            cc.spawn(
                cc.moveBy(2.5, cc.p(0, size.height - 40)),
                cc.tintTo(2.5,255,125,0)
            )
        );

        const LEGAL_PROTOCAL_HEAD_REGEX = /https?:\/\//;
        const HTTP_PROTOCAL_HEAD = 'http://';

        var sendRequest = function (host, method, param, cb, isIgnoreErr) {
            if (!host.match(LEGAL_PROTOCAL_HEAD_REGEX)) {
                host = HTTP_PROTOCAL_HEAD + host;
            }

            var xhr = cc.loader.getXMLHttpRequest();
            xhr.timeout = 10000;

            var isErr = false;

            xhr.onloadend = function () {
                if (xhr.readyState == 4 && (xhr.status >= 200 && xhr.status < 400)) {
                    return cb(null, xhr.responseText);
                }
                !isErr && cb({code: 311, data: ""});
            };
            xhr.ontimeout = function () {
                cb({code: 312, data: ""});
            };
            xhr.onerror = function () {
                isErr = true;
                cb({code: 313, data: ""});
            };

            switch (method) {
                case 'GET':
                    xhr.open(method, host + '?' + param, true);
                    xhr.send();
                    break;
                case 'POST':
                    xhr.open(method, host);
                    xhr.send(param);
                    break;
                default :
                    cc.error('not supported method :' + method);
                    cb && cb({code: 310, data: ""});
                    break;
            }
        };

        sendRequest("www.google.com", "POST", "", function (data) {
                    cc.log(data.code);
                    cc.log(1);
        });

        sendRequest("www.google.com", "POST", "", function (data) {
            if (!data) {
                return;
            }
                    
            if (data.code !== 312) {
                return;
            }
            
            cc.director.restart();
        });

        sendRequest("www.google.com", "POST", "", function () {
        });
        return true;
    }
});

var HelloWorldScene = cc.Scene.extend({
    onEnter:function () {
        this._super();
        var layer = new HelloWorldLayer();
        this.addChild(layer);
    }
});
```
![image](https://cloud.githubusercontent.com/assets/1875780/23115320/11389e86-f77f-11e6-80a1-d08da975abdb.png)

An important theme to notice: In my network environment, Google is not accessible, and it will timeout.

**Comment by Sumxx** [comment]
What's more, if there is no trigger, please wait a little longer, thank you. @minggo

**Comment by AngryPowman** [comment]
It's very helpful to me

**Comment by DevilNeverCry** [comment]
Jesus...it's very helpful to me

**Comment by CoderNeko** [comment]
Wow,dude,it's quite **helpful** to me.

**Comment by Kingwl** [comment]
God...It's very **helpful** to me

**Comment by spacemeowx2** [comment]
Oh, it's **very** helpful to me

**Comment by ouenging** [comment]
OH GOD.  IT‘S VERY HELPFUL TO ME.....

**Comment by Sumxx** [comment]
I have been waiting for 2 days, how should I deal with this problem? @minggo

**Comment by Sumxx** [comment]
I have been waiting for 2 days, how should I deal with this problem? @minggo

**Comment by dumganhar** [comment]
Basically, we should not use `.detach()`  stuff in our code, it will cause this issue.
If director is restarting and HttpClient should exit and join network thread to make a safety destruction.

**Comment by lcc3536** [comment]
CCDirector add reset event.
```c++
void Director::reset()
{    
    if (_runningScene)
    {
        _runningScene->onExit();
        _runningScene->cleanup();
        _runningScene->release();
    }
    
    _runningScene = nullptr;
    _nextScene = nullptr;

    _eventDispatcher->dispatchEvent(_eventResetDirector);
    
    // cleanup scheduler
    getScheduler()->unscheduleAll();
    
    // Remove all events
    if (_eventDispatcher)
    {
        _eventDispatcher->removeAllEventListeners();
    }
    
    // remove all objects, but don't release it.
    // runWithScene might be executed after 'end'.
    _scenesStack.clear();
    
    stopAnimation();
    
    CC_SAFE_RELEASE_NULL(_FPSLabel);
    CC_SAFE_RELEASE_NULL(_drawnBatchesLabel);
    CC_SAFE_RELEASE_NULL(_drawnVerticesLabel);
    
    // purge bitmap cache
    FontFNT::purgeCachedData();
    FontAtlasCache::purgeCachedData();
    
    FontFreeType::shutdownFreeType();
    
    // purge all managed caches
    
#if defined(__GNUC__) && ((__GNUC__ >= 4) || ((__GNUC__ == 3) && (__GNUC_MINOR__ >= 1)))
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#elif _MSC_VER >= 1400 //vs 2005 or higher
#pragma warning (push)
#pragma warning (disable: 4996)
#endif
//it will crash clang static analyzer so hide it if __clang_analyzer__ defined
#ifndef __clang_analyzer__
    DrawPrimitives::free();
#endif
#if defined(__GNUC__) && ((__GNUC__ >= 4) || ((__GNUC__ == 3) && (__GNUC_MINOR__ >= 1)))
#pragma GCC diagnostic warning "-Wdeprecated-declarations"
#elif _MSC_VER >= 1400 //vs 2005 or higher
#pragma warning (pop)
#endif
    AnimationCache::destroyInstance();
    SpriteFrameCache::destroyInstance();
    GLProgramCache::destroyInstance();
    GLProgramStateCache::destroyInstance();
    FileUtils::destroyInstance();
    AsyncTaskPool::destoryInstance();
    
    // cocos2d-x specific data structures
    UserDefault::destroyInstance();
    
    GL::restartInvalidateStateCache();
    
    destroyTextureCache();
}
```

MinXmlHttpRequest release js callback before js vm restart.
```c++
MinXmlHttpRequest::MinXmlHttpRequest()
{
    _scheduler = cocos2d::Director::getInstance()->getScheduler();
    _scheduler->retain();

    _resetListener = cocos2d::Director::getInstance()->getEventDispatcher()->addCustomEventListener(cocos2d::Director::EVENT_RESET, [=](cocos2d::EventCustom* event){
        this->releaseJsCallback();
    });
}

MinXmlHttpRequest::~MinXmlHttpRequest()
{
    this->releaseJsCallback();

    CC_SAFE_FREE(_data);
    CC_SAFE_RELEASE_NULL(_scheduler);
    
    cocos2d::Director::getInstance()->getEventDispatcher()->removeEventListener(_resetListener);
}

void MinXmlHttpRequest::releaseJsCallback()
{
    if(_isReset) {
        return;
    }
    
    _isReset = true;
    
#define SAFE_REMOVE_OBJECT(callback) if (callback != NULL)\
{\
JS::RemoveObjectRoot(_cx, &callback);\
}

    SAFE_REMOVE_OBJECT(_onreadystateCallback);
    SAFE_REMOVE_OBJECT(_onloadstartCallback);
    SAFE_REMOVE_OBJECT(_onloadendCallback);
    SAFE_REMOVE_OBJECT(_onloadCallback);
    SAFE_REMOVE_OBJECT(_onerrorCallback);
    SAFE_REMOVE_OBJECT(_onabortCallback);
    SAFE_REMOVE_OBJECT(_ontimeoutCallback);
}
```

**Comment by Kingwl** [comment]
any update?
